# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [1]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386975   30.664097   30.105382   30.415043  5.602586e+07
min      54.316940   55.316759   51.470005   55.215083  3.519590e+07
25%     112.581757  114.075474  111.125701  112.761704  7.636470e+07
50%     132.859970  133.991789  130.902101  132.681710  9.493580e+07
75%     148.053471  149.219241  146.111066  147.421466  1.296327e+08
max     178.443115  179.354932  175.609785  179.051010  4.265100e+08


In [2]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


/tmp/ipython-input-730687086.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386975   30.664097   30.105382   30.415043  5.602586e+07
min      54.316940   55.316759   51.470005   55.215083  3.519590e+07
25%     112.581757  114.075474  111.125701  112.761704  7.636470e+07
50%     132.859970  133.991789  130.902101  132.681710  9.493580e+07
75%     148.053471  149.219241  146.111066  147.421466  1.296327e+08
max     178.443115  179.354932  175.609785  179.051010  4.265100e+08


In [3]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipython-input-3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386975   30.664097   30.105382   30.415043  5.602586e+07
min      54.316940   55.316759   51.470005   55.215083  3.519590e+07
25%     112.581757  114.075474  111.125701  112.761704  7.636470e+07
50%     132.859970  133.991789  130.902101  132.681710  9.493580e+07
75%     148.053471  149.219241  146.111066  147.421466  1.296327e+08
max     178.443115  179.354932  175.609785  179.051010  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

In [4]:
# Suggest some features that might help improve a stock price prediction model.

# 1. Historical Prices: Include past closing, opening, high, and low prices.
# 2. Volume: Daily trading volumes.
# 3. Moving Averages: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
# 4. Technical Indicators: RSI, MACD, Bollinger Bands, etc.
# 5. Volatility: Historical volatility, Average True Range (ATR).
# 6. Sentiment Analysis: News articles, social media sentiment.
# 7. Economic Indicators: Interest rates, inflation rates, GDP growth.
# 8. Time-based Features: Day of the week, month, year.
# 9. Sector Performance: Comparison with industry and sector indices.
# 10. External Events: Major geopolitical events, earnings reports.

print("Suggested Features:")
print("1. Historical Prices: Include past closing, opening, high, and low prices.")
print("2. Volume: Daily trading volumes.")
print("3. Moving Averages: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).")
print("4. Technical Indicators: RSI, MACD, Bollinger Bands, etc.")
print("5. Volatility: Historical volatility, Average True Range (ATR).")
print("6. Sentiment Analysis: News articles, social media sentiment.")
print("7. Economic Indicators: Interest rates, inflation rates, GDP growth.")
print("8. Time-based Features: Day of the week, month, year.")
print("9. Sector Performance: Comparison with industry and sector indices.")
print("10. External Events: Major geopolitical events, earnings reports.")

Suggested Features:
1. Historical Prices: Include past closing, opening, high, and low prices.
2. Volume: Daily trading volumes.
3. Moving Averages: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. Technical Indicators: RSI, MACD, Bollinger Bands, etc.
5. Volatility: Historical volatility, Average True Range (ATR).
6. Sentiment Analysis: News articles, social media sentiment.
7. Economic Indicators: Interest rates, inflation rates, GDP growth.
8. Time-based Features: Day of the week, month, year.
9. Sector Performance: Comparison with industry and sector indices.
10. External Events: Major geopolitical events, earnings reports.


In [5]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Ensure the data is a pandas DataFrame with a DatetimeIndex
if not isinstance(data, pd.DataFrame):
    data = pd.DataFrame(data)
if not isinstance(data.index, pd.DatetimeIndex):
    data.index = pd.to_datetime(data.index)

# Drop the 'Price' column if it exists and is a duplicate of 'Close'
if 'Price' in data.columns and (data['Price'] == data['Close']).all():
    data = data.drop('Price', axis=1)

# Adopt the first four kinds of features: Historical Prices, Volume, Moving Averages, Technical Indicators
# For simplicity, we will focus on Historical Prices, Volume, and Moving Averages (5, 10, and 50 day).
# Technical Indicators can be more complex and are omitted for this example.

# Create Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Drop rows with NaN values created by moving averages
data.dropna(inplace=True)

# Define features (X) and target (y)
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
X = data[features]
y = data['Close'].shift(-1) # Predict tomorrow's closing price

# Drop the last row as the target is NaN
X = X[:-1]
y = y[:-1]

# Split data into training and testing sets (70/30 ratio)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Report performance metrics
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R²) Score: {r2}")

Mean Squared Error (MSE): 6.539002992755365
R-squared (R²) Score: 0.9918026189415472


In [6]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Create a binary target variable (1 if price increases tomorrow, 0 otherwise)
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Drop the last row as the target is NaN
data.dropna(inplace=True)

# Define features (X) and target (y) for classification
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
X_clf = data[features]
y_clf = data['Price_Increase']

# Split data into training and testing sets (70/30 ratio)
X_train_clf, X_test_clf, y_train_clf, y_test_clf = train_test_split(X_clf, y_clf, test_size=0.3, random_state=42)

# Train a Random Forest Classifier model
model_clf = RandomForestClassifier(n_estimators=100, random_state=42)
model_clf.fit(X_train_clf, y_train_clf)

# Make predictions
y_pred_clf = model_clf.predict(X_test_clf)

# Report performance metrics
accuracy = accuracy_score(y_test_clf, y_pred_clf)
conf_matrix = confusion_matrix(y_test_clf, y_pred_clf)
class_report = classification_report(y_test_clf, y_pred_clf)

print(f"Accuracy: {accuracy}")
print(f"Confusion Matrix:\n{conf_matrix}")
print(f"Classification Report:\n{class_report}")

# Report Feature Importance
feature_importances = pd.Series(model_clf.feature_importances_, index=features).sort_values(ascending=False)
print(f"\nFeature Importance:\n{feature_importances}")

Accuracy: 0.45539906103286387
Confusion Matrix:
[[43 61]
 [55 54]]
Classification Report:
              precision    recall  f1-score   support

           0       0.44      0.41      0.43       104
           1       0.47      0.50      0.48       109

    accuracy                           0.46       213
   macro avg       0.45      0.45      0.45       213
weighted avg       0.45      0.46      0.45       213


Feature Importance:
High      0.130715
Open      0.129279
MA_10     0.127711
Volume    0.125547
MA_5      0.125515
MA_50     0.122988
Close     0.119873
Low       0.118372
dtype: float64


In [7]:
# prompt: Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

# Download historical data for AAPL for the new period
new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Ensure the new data is a pandas DataFrame with a DatetimeIndex
if not isinstance(new_data, pd.DataFrame):
    new_data = pd.DataFrame(new_data)
if not isinstance(new_data.index, pd.DatetimeIndex):
    new_data.index = pd.to_datetime(new_data.index)

# Drop the 'Price' column if it exists and is a duplicate of 'Close'
if 'Price' in new_data.columns and (new_data['Price'] == new_data['Close']).all():
    new_data = new_data.drop('Price', axis=1)

# Create Moving Averages for the new data
new_data['MA_5'] = new_data['Close'].rolling(window=5).mean()
new_data['MA_10'] = new_data['Close'].rolling(window=10).mean()
new_data['MA_50'] = new_data['Close'].rolling(window=50).mean()

# Create the binary target variable for the new data
new_data['Price_Increase'] = (new_data['Close'].shift(-1) > new_data['Close']).astype(int)

# Drop rows with NaN values created by moving averages and the target variable
new_data.dropna(inplace=True)

# Define features (X_new) and target (y_new) for the new data
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
X_new = new_data[features]
y_new = new_data['Price_Increase']

# Make predictions using the trained Random Forest Classifier model (model_clf)
y_pred_new = model_clf.predict(X_new)

# Report performance metrics on the new data
accuracy_new = accuracy_score(y_new, y_pred_new)
conf_matrix_new = confusion_matrix(y_new, y_pred_new)
class_report_new = classification_report(y_new, y_pred_new)

print(f"Accuracy on new data: {accuracy_new}")
print(f"Confusion Matrix on new data:\n{conf_matrix_new}")
print(f"Classification Report on new data:\n{class_report_new}")

/tmp/ipython-input-2971233571.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed


Accuracy on new data: 0.472636815920398
Confusion Matrix on new data:
[[ 87   5]
 [101   8]]
Classification Report on new data:
              precision    recall  f1-score   support

           0       0.46      0.95      0.62        92
           1       0.62      0.07      0.13       109

    accuracy                           0.47       201
   macro avg       0.54      0.51      0.38       201
weighted avg       0.55      0.47      0.36       201



# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



Based on the performance metrics reported in the previous cell:

- **Accuracy:** The accuracy on the new data is around 47.26%, which is slightly better than random guessing (50%).
- **Confusion Matrix:**